discovery/kubernetes/client_metrics.go

```diff
@@ -0,0 +1,225 @@
+			Name:      "http_request_result_total",
+			Help:      "Request results of the client-go on the Kubernetes API.",
+		},
+		[]string{"status_code", "method"},
+	)
+	clientGoRequestLatencyMetric = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace: metricsNamesapce,
+			Name:      "http_request_duration_seconds",
+			Help:      "Request latency of the client-go on Kubernetes API.",
+		},
+		[]string{"verb", "endpoint"},
+	)
+
+	// Definition of metrics for client-go cache metrics provider
+	clientGoCacheListTotalMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "cache_list_total",
+			Help:      "Total number of client-go Kubernets API calls.",
+		},
+	)
+	clientGoCacheListDurationMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_duration_seconds",
+			Help:      "Duration of a Kubernetes API call in seconds.",
+		},
+	)
+	clientGoCacheItemsInListCountMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_items_count",
+			Help:      "Count of items in a list from the Kubernetes API.",
+		},
+	)
+	clientGoCacheWatchesCountMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watches_total",
+			Help:      "Count of watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheShortWatchesCountMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "short_watches_total",
+			Help:      "Count of short watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheWatchesDurationMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watch_duration_seconds",
+			Help:      "Duration of watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheItemsInWatchesCountMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watch_items",
+			Help:      "Number of items in watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheLastResourceVersionMetric = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "last_resource_version",
+			Help:      "Last resource version from the Kubernetes API.",
+		},
+	)
+
+	// Definition of metrics for client-go workflow metrics provider
+	clientGoWorkqueueDepthMetric = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: workqueueMetricsNamespace,
+			Name:      "depth",
+			Help:      "Current depth of a client-go workqueue.",
+		},
+	)
+	clientGoWorkqueueAddsMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: workqueueMetricsNamespace,
+			Name:      "adds_total",
+			Help:      "Total count of adds handeled by a client-go workqueue.",
+		},
+	)
+	clientGoWorkqueueRetriesMetric = prometheus.NewCounter(
```

Comment:
IIUC only queues created by `workqueue.NewNamedDelayingQueue()` has support for retries and we're using `workqueue.NewNamedQueue()` currently so this metric will always be 0.